src/nhs_grid.jl

```diff
@@ -154,40 +154,45 @@ function update_grid!(neighborhood_search::GridNeighborhoodSearch, coords_fun)
         for cell_index in update_buffer[j]
             points = cell_list[cell_index]
 
-            # Find all points whose coordinates do not match this cell
-            moved_point_indices = (i for i in eachindex(points)
-                                   if !is_correct_cell(cell_list,
-                                                       cell_coords(coords_fun(points[i]),
-                                                                   neighborhood_search),
-                                                       cell_index))
-
-            # Add moved points to new cell
-            for i in moved_point_indices
+            # Find all points whose coordinates do not match this cell.
+            #
+            # WARNING!!!
+            # The `DynamicVectorOfVectors` requires this loop to be **in descending order**.
+            # `deleteat_cell!(..., i)` will change the order of points that come after `i`.
+            for i in reverse(eachindex(points))
                 point = points[i]
-                new_cell_coords = cell_coords(coords_fun(point), neighborhood_search)
+                cell_coords_ = cell_coords(coords_fun(point), neighborhood_search)
 
-                # Add point to corresponding cell or create cell if it does not exist
-                push_cell!(cell_list, new_cell_coords, point)
-            end
+                if !is_correct_cell(cell_list, cell_coords_, cell_index)
+                    new_cell_coords = cell_coords(coords_fun(point), neighborhood_search)
 
-            # Remove moved points from this cell
-            deleteat_cell!(cell_list, cell_index, moved_point_indices)
+                    # Add point to new cell or create cell if it does not exist
+                    push_cell!(cell_list, new_cell_coords, point)
+
+                    # Remove moved point from this cell
+                    deleteat_cell!(cell_list, cell_index, i)
+                end
+            end
```

Comment:
This new update is actually faster than the old generator bullsh*t.
5-10% faster in 2D and ~50% faster in 3D. Here are benchmarks with #41:
```
julia> plot_benchmarks(benchmark_update_alternating, (20, 20), 6, title="Original Update 2D")
GridNeighborhoodSearch
with 20x20 = 400 particles finished in 7.185 μs

GridNeighborhoodSearch
with 40x40 = 1600 particles finished in 14.376 μs

GridNeighborhoodSearch
with 80x80 = 6400 particles finished in 41.341 μs

GridNeighborhoodSearch
with 160x160 = 25600 particles finished in 122.243 μs

GridNeighborhoodSearch
with 320x320 = 102400 particles finished in 546.118 μs

GridNeighborhoodSearch
with 640x640 = 409600 particles finished in 2.319 ms

julia> plot_benchmarks(benchmark_update_alternating, (20, 20), 6, title="New Update 2D")
GridNeighborhoodSearch
with 20x20 = 400 particles finished in 5.280 μs

GridNeighborhoodSearch
with 40x40 = 1600 particles finished in 11.375 μs

GridNeighborhoodSearch
with 80x80 = 6400 particles finished in 31.820 μs

GridNeighborhoodSearch
with 160x160 = 25600 particles finished in 108.847 μs

GridNeighborhoodSearch
with 320x320 = 102400 particles finished in 491.897 μs

GridNeighborhoodSearch
with 640x640 = 409600 particles finished in 2.208 ms

julia> plot_benchmarks(benchmark_update_alternating, (7.37, 7.37, 7.37), 6, title="Original Update 3D")
GridNeighborhoodSearch
with 7x7x7 = 343 particles finished in 2.826 μs

GridNeighborhoodSearch
with 12x12x12 = 1728 particles finished in 22.925 μs

GridNeighborhoodSearch
with 19x19x19 = 6859 particles finished in 76.967 μs

GridNeighborhoodSearch
with 29x29x29 = 24389 particles finished in 277.532 μs

GridNeighborhoodSearch
with 47x47x47 = 103823 particles finished in 1.236 ms

GridNeighborhoodSearch
with 74x74x74 = 405224 particles finished in 5.510 ms

julia> plot_benchmarks(benchmark_update_alternating, (7.37, 7.37, 7.37), 6, title="New Update 3D")
GridNeighborhoodSearch
with 7x7x7 = 343 particles finished in 2.821 μs

GridNeighborhoodSearch
with 12x12x12 = 1728 particles finished in 15.415 μs

GridNeighborhoodSearch
with 19x19x19 = 6859 particles finished in 46.851 μs

GridNeighborhoodSearch
with 29x29x29 = 24389 particles finished in 156.888 μs

GridNeighborhoodSearch
with 47x47x47 = 103823 particles finished in 654.662 μs

GridNeighborhoodSearch
with 74x74x74 = 405224 particles finished in 3.177 ms
```